book-content/chapters/16-the-utils-folder.md

```diff
@@ -0,0 +1,1527 @@
+    if (Math.random() > 0.5) {
+      throw new Error("Something went wrong");
+    }
+    return 123;
+  });
+  ...
+
+// hovering over safeFunction shows:
+const safeFunction: <number>(func: PromiseFunc<number>) => Promise<() => Promise<number | Error>>
+```
+
+The other tests pass as well.
+
+Whatever we pass into `safeFunction` will be inferred as the type argument for `PromiseFunc`. This is because the type argument is being inferred _inside_ the generic function.
+
+This combination of generic functions and generic types can make your generic functions a lot easier to read.
+
+### Solution 6: Multiple Type Arguments in a Generic Function
+
+Here's how `PromiseFunc` is currently defined:
+
+```typescript
+type PromiseFunc<TResult> = (...args: any[]) => Promise<TResult>;
+```
+
+The first thing to do is figure out the types of the arguments being passed in. Currently, they're set to one value, but they need to be different based on the type of function being passed in.
+
+Instead of having `args` be of type `any[]`, we want to spread in all of the `args` and capture the entire array.
+
+To do this, we'll update the type to be `TArgs`. Since `args` needs to be an array, we'll say that `TArgs extends any[]`. Note that this doesn't mean that `TArgs` will be typed as `any`, but rather that it will accept any kind of array:
+
+```typescript
+type PromiseFunc<TArgs extends any[], TResult> = (
+  ...args: TArgs
+) => Promise<TResult>;
+```
+
+You might have tried this with `unknown[]` - but `any[]` is the only thing that works in this scenario.
+
+Now we need to update the `safeFunction` so that it has the same arguments as `PromiseFunc`. To do this, we'll add `TArgs` to its type parameters.
+
+Note that we also need to update the args for the `async` function to be of type `TArgs`:
+
+```typescript
+const safeFunction =
+  <TArgs extends any[], TResult>(func: PromiseFunc<TArgs, TResult>) =>
+  async (...args: TArgs) => {
+    try {
+      const result = await func(...args);
+      return result;
+    } catch (e) {
+      ...
+```
+
+This change is necessary in order to make sure the function returned by `safeFunction` has the same typed arguments as the original function.
+
+With these changes, all of our tests pass as expected.
+
+### Solution 8: Assertion Functions
+
+The solution is to add a type annotation onto the return type of `assertIsAdminUser`.
+
+If it was a type predicate, we would say `user is AdminUser`:
+
+```ts twoslash
+// @errors: 2355
+type User = {
+  id: string;
+  name: string;
+};
+type AdminUser = {
+  id: string;
+  name: string;
+  roles: string[];
+};
+
+// ---cut---
+function assertIsAdminUser(user: User): user is AdminUser {
+  if (!("roles" in user)) {
+    throw new Error("User is not an admin");
+  }
+}
+```
+
+However, this leads to an error. We get this error because `assertIsAdminUser` is returning `void`, which is different from a type predicate that requires you to return a Boolean.
```

Comment:
```suggestion
However, this leads to an error. We get this error because `assertIsAdminUser` is returning `void`, which is different from a type predicate that requires you to return a boolean.
```